File: backend/models/invoice.model.js
module.exports = (sequelize, DataTypes) => {
  const Invoice = sequelize.define('Invoice', {
    invoice_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    owner_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: 'pet_owners',
        key: 'owner_id'
      }
    },
    appointment_id: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: {
        model: 'appointments',
        key: 'appointment_id'
      }
    },
    invoice_number: {
      type: DataTypes.STRING(50),
      allowNull: false,
      unique: true
    },
    subtotal: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      defaultValue: 0,
      validate: {
        min: { args: 0, msg: 'Subtotal cannot be negative' }
      }
    },
    tax: {
      type: DataTypes.DECIMAL(10, 2),
      defaultValue: 0,
      validate: {
        min: { args: 0, msg: 'Tax cannot be negative' }
      }
    },
    discount: {
      type: DataTypes.DECIMAL(10, 2),
      defaultValue: 0,
      validate: {
        min: { args: 0, msg: 'Discount cannot be negative' }
      }
    },
    total_amount: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      validate: {
        min: { args: 0, msg: 'Total amount cannot be negative' }
      }
    },
    payment_status: {
      type: DataTypes.ENUM('Pending', 'Paid', 'Partially Paid', 'Cancelled', 'Refunded'),
      defaultValue: 'Pending'
    },
    payment_method: {
      type: DataTypes.ENUM('Cash', 'Card', 'Bank Transfer', 'Online'),
      allowNull: true
    },
    invoice_date: {
      type: DataTypes.DATEONLY,
      allowNull: false,
      validate: {
        isDate: { msg: 'Invalid date format' }
      }
    },
    due_date: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    paid_date: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    paid_amount: {
      type: DataTypes.DECIMAL(10, 2),
      defaultValue: 0,
      validate: {
        min: { args: 0, msg: 'Paid amount cannot be negative' }
      }
    },
    notes: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    created_by: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: {
        model: 'users',
        key: 'user_id'
      }
    },
    created_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW
    },
    updated_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW
    }
  }, {
    tableName: 'invoices',
    timestamps: false,
    indexes: [
      { fields: ['owner_id'] },
      { fields: ['invoice_number'] },
      { fields: ['payment_status'] },
      { fields: ['invoice_date'] }
    ]
  });

  // Define associations
  Invoice.associate = (models) => {
    Invoice.belongsTo(models.PetOwner, {
      foreignKey: 'owner_id',
      as: 'owner'
    });

    Invoice.belongsTo(models.Appointment, {
      foreignKey: 'appointment_id',
      as: 'appointment'
    });

    Invoice.belongsTo(models.User, {
      foreignKey: 'created_by',
      as: 'creator'
    });

    Invoice.hasMany(models.InvoiceItem, {
      foreignKey: 'invoice_id',
      as: 'items'
    });
  };

  return Invoice;
};

File: mobile_app/lib/models/invoice_model.dart
class InvoiceModel {
  final int    invoiceId;
  final int    ownerId;
  final String invoiceNumber;
  final double subtotal;
  final double taxRate;
  final double taxAmount;
  final double discount;
  final double totalAmount;
  final String paymentMethod;
  final String paymentStatus;
  final String? notes;
  final String? dueDate;
  final String  invoiceDate;
  final String? paidDate;
  final String? ownerName;
  final String? ownerPhone;
  final String? serviceType;
  final String? branchName;
  final List<InvoiceItemModel> items;

  InvoiceModel({
    required this.invoiceId,
    required this.ownerId,
    required this.invoiceNumber,
    required this.subtotal,
    required this.taxRate,
    required this.taxAmount,
    required this.discount,
    required this.totalAmount,
    required this.paymentMethod,
    required this.paymentStatus,
    this.notes,
    this.dueDate,
    required this.invoiceDate,
    this.paidDate,
    this.ownerName,
    this.ownerPhone,
    this.serviceType,
    this.branchName,
    this.items = const [],
  });

  factory InvoiceModel.fromJson(Map<String, dynamic> json) {
    return InvoiceModel(
      invoiceId:     json['invoice_id']     ?? 0,
      ownerId:       json['owner_id']       ?? 0,
      invoiceNumber: json['invoice_number'] ?? '',
      subtotal:      double.tryParse(
        json['subtotal'].toString()) ?? 0,
      taxRate:       double.tryParse(
        json['tax_rate'].toString()) ?? 0,
      taxAmount:     double.tryParse(
        json['tax_amount'].toString()) ?? 0,
      discount:      double.tryParse(
        json['discount'].toString()) ?? 0,
      totalAmount:   double.tryParse(
        json['total_amount'].toString()) ?? 0,
      paymentMethod: json['payment_method'] ?? 'Cash',
      paymentStatus: json['payment_status'] ?? 'Pending',
      notes:         json['notes'],
      dueDate:       json['due_date'],
      invoiceDate:   json['invoice_date']   ?? '',
      paidDate:      json['paid_date'],
      ownerName:     json['owner_name'],
      ownerPhone:    json['owner_phone'],
      serviceType:   json['service_type'],
      branchName:    json['branch_name'],
      items: (json['items'] as List<dynamic>? ?? [])
          .map((i) => InvoiceItemModel.fromJson(i))
          .toList(),
    );
  }

  bool get isPaid    => paymentStatus == 'Paid';
  bool get isPending => paymentStatus == 'Pending';
}

class InvoiceItemModel {
  final int    itemId;
  final String description;
  final double quantity;
  final double unitPrice;
  final double totalPrice;

  InvoiceItemModel({
    required this.itemId,
    required this.description,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
  });

  factory InvoiceItemModel.fromJson(Map<String, dynamic> json) {
    return InvoiceItemModel(
      itemId:      json['item_id']     ?? 0,
      description: json['description'] ?? '',
      quantity:    double.tryParse(
        json['quantity'].toString()) ?? 0,
      unitPrice:   double.tryParse(
        json['unit_price'].toString()) ?? 0,
      totalPrice:  double.tryParse(
        json['total_price'].toString()) ?? 0,
    );
  }
}